Prevent UDN deletion race condition by checking controller state

kyrtapz opened the pull request:
Adds `GetActiveNetwork` method to verify network controllers are stopped before allowing UDN/CUDN deletion, preventing race conditions in a scenario where a network is recreated before the cluster-manager controller stopped running.
Replaces custom condition update logic with standard k8s `meta.SetStatusCondition` to avoid updating condtions even if they haven't changed. Use rate limiting instead of fixed interval re-queuing. 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Prevent deletion of user-defined network resources while their network controllers are active.
  * Added ability to query whether a network controller is active.
  * Use rate-limited retries during reconciliation and modernized status condition handling.

* **Bug Fixes**
  * More reliable removal of network resources by detecting active controllers.
  * Consistent, sorted pod lists in error messages.

* **Tests**
  * Tests updated to use the network-manager interface for controller scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kyrtapz commented:
/retest-failed

jcaamano commented on go-controller/pkg/clustermanager/userdefinednetwork/controller.go:
```diff
@@ -410,6 +415,10 @@ func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefine
 				return nil, fmt.Errorf("failed to delete NetworkAttachmentDefinition [%s/%s]: %w", udn.Namespace, udn.Name, err)
 			}
 
+			if c.networkManager.GetControllerNetwork(util.GenerateUDNNetworkName(udn.Namespace, udn.Name)) != nil {
```
Do you think it would do good to have comments here. This can be a hotspot where people might look at it an ask "why are we waiting?"

kyrtapz replied:
Great idea, will do.

jcaamano commented on go-controller/pkg/clustermanager/userdefinednetwork/controller.go:
```diff
@@ -389,7 +387,7 @@ func (c *Controller) reconcileUDN(key string) error {
 	var networkInUse *networkInUseError
 	if errors.As(syncErr, &networkInUse) {
-		c.udnController.ReconcileAfter(key, c.networkInUseRequeueInterval)
+		c.udnController.ReconcileRateLimited(key)
```
While the change seems reasonable, I am missing if there was an actual need to make the change or not. If not, did you check for all other uses of `ReconcileAfter` if they could reasonable be changed as well? I wouldn't like for `ReconcileAfter` to remain there if it is most likely an improper use most of the time.

jcaamano replied:
> I am not a fan of not doing rate limited enqueuing by default

+1

> The only other component using `ReconcileAfter` is the route_import controller and I would like to hear from you whether it is desired there.

For of its uses it is reasonable: when we receive a route update we will potentially receive more route updates immediately after that one so I wait a bit to reconcile.
For all its other uses, it falls in the category of your other comment that reconciles should be rate limited by default.

I guess no action on this front then for now.

> Without this change it can be the case that in normal operation the controller is still running when the UDN delete is processed which would cause a huge delay of 1 minute.

Not sure I fully understand.
`defaultNetworkInUseCheckInterval` is 1 minute so you are saying that is too much of a delay?
If you use the queue rate limiter the requeue can be immediate with no delay depending on what the controller might be doing so I guess that's the drawback?
Have you considered reducing `defaultNetworkInUseCheckInterval`?

jcaamano replied:
> Only initially, if the reconciliation fails then the delay increases. Not sure what is the drawback you have in mind.

The drawback of retying before it can be reasonably expected for whatever cause to be to holding off the delete of the netwoprk to have cleared up. From the initial rate limit ~10ms to 1min there is a big gap.

> My understanding was that our plain reconciliation isn't rate limited?

A retry for an error is rate limited:
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/68bd339b032cdd2881cf5fa3951c8dc2b04494f7/go-controller/pkg/controller/controller.go#L258

So just returning the error would have the same effect.

kyrtapz replied:
> A retry for an error is rate limited:

I've missed that initially, sorry. I don't think the behavior is the same in this case since this controller(as most of our controllers) stops retrying after 15 tries. For this specific error we want to keep retrying with the hopes that:
- The controller stopped(very likely)
- The pods were eventually removed(less likely and can take a long time).

jcaamano replied:
If it were me I would just set the controller config max attempts to `InfiniteAttempts`.

But I hear you and this now becomes a nit, so I am ok with your decision and hope that if we ever do it this doesn't go missed to be cleaned up, even if harmless.

jcaamano commented on go-controller/pkg/clustermanager/userdefinednetwork/controller.go:
```diff
@@ -450,40 +448,40 @@ func (c *Controller) updateUserDefinedNetworkStatus(udn *userdefinednetworkv1.Us
+		return fmt.Errorf("failed to update UserDefinedNetwork status: %w", err)
 	}
-
+	klog.Infof("Updated status UserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
```
noop comment: this triggers my "excessive logging" thought

jcaamano replied:
with noop I mean I don't expect you to do anything, just a thought of mine.

However in regards to your comment, I think we should just make proper use of levels and be able to reproduce the issues. Logging on every update is too much logging in my view. If we did that everywhere it would be bad.

jcaamano commented on go-controller/pkg/networkmanager/api.go:
```diff
@@ -42,6 +42,11 @@ type Interface interface {
+	// GetControllerNetwork returns the netinfo currently held by the controller for the given network.
+	// This may differ from the netinfo returned by GetNetwork which reflects the API state.
+	// Returns nil if there is no running controller for the provided network.
+	GetControllerNetwork(network string) util.NetInfo
```
I can't outright understand what it tries to say. What is `Controller` referring to? We are not getting any controller.

jcaamano replied:
> network used by the controller

hmmm what controller? I am just missing to what controller you a referring to. Do you mean the one in `network_controller.go` in the package? I see that as an implementation detail of the package, the caller is just calling network manager, and it has no idea there is a controller in there or what can that possibly mean towards the functionality of the method. It is only meaningful if you are aware how the package is implemented and I had to think hard :p

`GetNetworkState` does look good to me, because if there is state it kinda means the network is active, and if not, it means the network is not active. But I can give you other ideas in order of preference: `GetActiveNetwork`, `GetActiveNetworkState`, `GetRunningNetworkState`, `GetRunningNetwork` ....

kyrtapz replied:
I think that `GetControllerNetwork` is less abstract but more clear to ovn-k developers. Going with `GetNetworkState`(a method of an interface called `Interface`) makes the code more abstract but less clear and can be kinda understood depending on who looks at it. If the abstraction is the main focus of this interface I can change it to `GetNetworkState` as it is vague enough to be interpreted as anything - I am not sure what a network state really is.

jcaamano commented on go-controller/pkg/clustermanager/userdefinednetwork/controller.go:
```diff
@@ -450,40 +448,40 @@ func (c *Controller) updateUserDefinedNetworkStatus(udn *userdefinednetworkv1.Us
 }
 
-func newNetworkCreatedCondition(nad *netv1.NetworkAttachmentDefinition, syncError error) *metav1.Condition {
+func newNetworkCreatedCondition(nad *netv1.NetworkAttachmentDefinition, syncError error) metav1.Condition {
```
up to you, just a nit.

kyrtapz commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/aea1e27b7406fb7a362f02f1d87abf6c9675a364..2f1987c623d208e0b403b8e6aa77a8055eb38bbb):
- Added some comments
- Renamed GetControllerNetwork to GetActiveNetwork and moved it right under GetNetwork
- Reverted back to returning a pointer in newNetworkCreatedCondition.

jcaamano commented on go-controller/pkg/clustermanager/userdefinednetwork/controller.go:
```diff
@@ -410,6 +414,12 @@ func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefine
 				return nil, fmt.Errorf("failed to delete NetworkAttachmentDefinition [%s/%s]: %w", udn.Namespace, udn.Name, err)
 			}
 
+			// Ensure that the network controller is stopped(GetActiveNetwork returns nil) before allowing the UDN
+			// to be removed.
```
Can you enrich this comment a bit. I will make a proposal.
```
This allows for annotations of the old network to be removed before a potential new network with the same name is added preventing the new network to use the old annotations on the node side
```
